# 102 Python intermediate - lambda, map, filter, comprehensions
_Kamil Bartocha_

_wersja_ 0.0.1

## Comprehensions

W Pythonie **comprehensions** (wyrażenia listowe, słownikowe, zbiorów i generatorów) to elegancka i zwięzła konstrukcja do tworzenia nowych kolekcji na podstawie istniejących w sposób przejrzysty i czytelny. Pozwalają one tworzyć listy, słowniki, zbiory lub generatory za pomocą pojedynczej linii kodu, często z filtrowaniem i transformacją danych.

### Zalety comprehensions:
**Zwięzłość**: Pozwalają na zapisanie kodu w jednej linii, co zwiększa czytelność.

**Wydajność**: W przypadku generatorów oszczędzają pamięć, generując elementy na bieżąco.

**Czytelność**: Często są bardziej zrozumiałe niż rozbudowane pętle for.

### 1. **List comprehensions (wyrażenia listowe)**

List comprehensions pozwalają na wygenerowanie nowej listy na podstawie istniejącej iterowalnej struktury (np. innej listy, zakresu). Składnia jest następująca:

```python
new_list = [expression for item in iterable if condition]
```
gdzie:

- `expression` – dowolna operacja wykonywana na każdym elemencie
- `item` – bieżący element w iteracji
- `iterable` – struktura danych, którą iterujemy (np. lista, range)
- `condition` – (opcjonalnie) warunek, który musi być spełniony, aby dodać element do nowej listy

#### Przykłady:

In [1]:
squares = [x**2 for x in range(10)]
print(squares)


[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


In [2]:
even_numbers = [x for x in range(10) if x % 2 == 0]
print(even_numbers)

[0, 2, 4, 6, 8]


In [3]:
words = ["hello", "world", "python"]
upper_words = [word.upper() for word in words]
print(upper_words)


['HELLO', 'WORLD', 'PYTHON']


### 2. **Dictionary comprehensions (wyrażenia słownikowe)**

Podobnie jak w przypadku list comprehensions, można wygenerować nowy słownik, iterując po elementach iterowalnych i przypisując klucz oraz wartość dla każdego elementu.

```python
new_dict = {key_expression: value_expression for item in iterable if condition}
```
gdzie:

- `key_expression` – klucz w nowym elemencie
- `value_expression` – wartość w nowym elemencie
- `item` – bieżący element w iteracji
- `iterable` – struktura danych, którą iterujemy (np. lista, range)
- `condition` – (opcjonalnie) warunek, który musi być spełniony, aby dodać element do nowej listy

#### Przykłady:

In [4]:
squares_dict = {x: x**2 for x in range(5)}
print(squares_dict)

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16}


In [6]:
words = ['apple', 'banana', 'cherry', 'ananas']
length_dict = {word: len(word) for word in words if word.startswith('a')}
print(length_dict)

{'apple': 5, 'ananas': 6}


### 3. **Set comprehensions (wyrażenia zbiorów)**

Set comprehensions są podobne do list comprehensions, ale tworzą zbiór (unikalną kolekcję elementów).

```python
new_set = {expression for item in iterable if condition}
```
gdzie:

- `expression` – wyrażenie w nowym elemencie
- `item` – bieżący element w iteracji
- `iterable` – struktura danych, którą iterujemy (np. lista, range)
- `condition` – (opcjonalnie) warunek, który musi być spełniony, aby dodać element do nowej listy

#### Przykłady:

In [9]:
squares_set = {x ** 2 for x in range(10)}
print(squares_set)


{0, 1, 64, 4, 36, 9, 16, 49, 81, 25}


In [8]:
squares_set_odds = {x ** 2 for x in range(10) if x % 2 == 1}
print(squares_set_odds)

{1, 9, 81, 49, 25}


### 4. **Generator comprehensions (wyrażenia generatorów)**

Wyrażenia generatorów są podobne do list comprehensions, ale zamiast tworzyć listę w pamięci, zwracają obiekt generatora, który generuje elementy "na żądanie", oszczędzając pamięć.

```python
generator = (expression for item in iterable if condition)
```
gdzie:

- `expression` – wyrażenie w nowym elemencie
- `item` – bieżący element w iteracji
- `iterable` – struktura danych, którą iterujemy (np. lista, range)
- `condition` – (opcjonalnie) warunek, który musi być spełniony, aby dodać element do nowej listy

#### Przykłady:

In [16]:
squares_gen = (x**2 for x in range(10))
print(squares_gen)
for square in squares_gen:
    print(square)

squares_gen2 = (x**2 for x in range(10))
print(list(squares_gen2))

<generator object <genexpr> at 0x110589a80>
0
1
4
9
16
25
36
49
64
81
[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]


**Różnica między list comprehension a generator comprehension**

Główna różnica polega na tym, że list comprehension tworzy całą listę w pamięci od razu, podczas gdy generator comprehension generuje elementy w locie, co pozwala na oszczędność zasobów. Oto prosty przykład porównania:


**Główne zalety generatorów:**

- *Wydajność pamięci:* Idealne w przypadku pracy z bardzo dużymi zbiorami danych, ponieważ generator nie trzyma całej kolekcji w pamięci.

- *Lenistwo (lazy evaluation):* Wartości są generowane dopiero wtedy, kiedy są potrzebne, co jest wydajne, gdy nie chcemy lub nie potrzebujemy wszystkich wyników od razu.

- *Jednokrotne przechodzenie:* W odróżnieniu od list, które można przechodzić wielokrotnie, generatory pozwalają na jednokrotne przejście przez elementy.


In [17]:
# List comprehension (wszystkie elementy są przechowywane w pamięci)
squares_list = [x**2 for x in range(1000)]

# Generator comprehension (elementy są generowane na bieżąco)
squares_gen = (x**2 for x in range(1000))

print(squares_list)
print(squares_gen)

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801, 10000, 10201, 10404, 10609, 10816, 11025, 11236, 11449, 11664, 11881, 12100, 12321, 12544, 12769, 12996, 13225, 13456, 13689, 13924, 14161, 14400, 14641, 14884, 15129, 15376, 15625, 15876, 16129, 16384, 16641, 16900, 17161, 17424, 17689, 17956, 18225, 18496, 18769, 19044, 19321, 19600, 19881, 20164, 20449, 20736, 21025, 21316, 21609, 21904, 22201, 22500, 22801, 23104, 23409, 23716, 24025, 24336, 24649, 24964, 25281, 25600, 25921, 26244, 2656

#### Przykład zaawansowany:

Załóżmy, że chcemy generować nieskończoną liczbę liczb całkowitych. Generator będzie idealnym narzędziem do takiej operacji:

In [2]:
def infinite_numbers(start=0):
    while True:
        yield start
        start += 1

gen = infinite_numbers()
for i in range(5):
    print(next(gen))  # Zwraca kolejne liczby, zaczynając od 0


gen = infinite_numbers()
for i in range(5):
    print(next(gen))

0
1
2
3
4
0
1
2
3
4


#### **Generator a `yield`**

Podobnie jak w przypadku generatorów tworzonych za pomocą comprehensions, w Pythonie możemy także tworzyć generatory za pomocą funkcji z użyciem słowa kluczowego `yield`. Funkcja z `yield` zwraca generator.

#### Przykład funkcji z yield:

In [4]:
def count_up_to(max_value):
    count = 1
    while count <= max_value:
        yield count   # return count i stop funkcja
        count += 1

gen = count_up_to(5)
print(list(gen))

[1, 2, 3, 4, 5]


`yield` generują wartości sekwencyjnie, pozwalając na leniwą (lazy) ewaluację. Gdy funkcja zawiera yield, zamiast zwykłego zwracania wyniku za pomocą `return`, staje się ona generatorem.

Funkcja z `yield` działa inaczej niż zwykła funkcja – nie wykonuje się w całości od razu, lecz wstrzymuje (pauzuje) swoje działanie za każdym razem, gdy napotka yield, zwracając bieżącą wartość, a przy kolejnym wywołaniu kontynuuje działanie od miejsca, gdzie została wstrzymana.

Gdy wywołujemy tę funkcję, nie zwraca ona od razu wartości jak `return`, lecz zwraca obiekt generatora.
Kiedy użyjemy funkcji takich jak `next()`, generator "wyciąga" kolejną wartość wygenerowaną przez `yield`.

In [6]:
gen = count_up_to(3)

print(next(gen))  # Output: 1
print(next(gen))  # Output: 2
print(next(gen))  # Output: 3
print(next(gen))  # Output: StopIteration bo max=3



1
2
3


StopIteration: 

Po ostatnim `yield`, funkcja kończy działanie, a kolejne wywołanie `next()` podniesie wyjątek `StopIteration`.

- `return`: Kończy działanie funkcji i zwraca wartość od razu.
- `yield`: Zawiesza działanie funkcji, zachowując jej stan (wartości zmiennych lokalnych), pozwalając na późniejsze wznowienie jej wykonania.

#### Przykład zaawansowany: Fibbonacci

In [8]:
def fibonacci():
    a, b = 0, 1
    while True:
        yield a
        a, b = b, a + b


fib_gen = fibonacci()
x = 1_000

for _ in range(x):
    print(next(fib_gen))

print('*' * 20)
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))
print(next(fib_gen))




0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181
6765
10946
17711
28657
46368
75025
121393
196418
317811
514229
832040
1346269
2178309
3524578
5702887
9227465
14930352
24157817
39088169
63245986
102334155
165580141
267914296
433494437
701408733
1134903170
1836311903
2971215073
4807526976
7778742049
12586269025
20365011074
32951280099
53316291173
86267571272
139583862445
225851433717
365435296162
591286729879
956722026041
1548008755920
2504730781961
4052739537881
6557470319842
10610209857723
17167680177565
27777890035288
44945570212853
72723460248141
117669030460994
190392490709135
308061521170129
498454011879264
806515533049393
1304969544928657
2111485077978050
3416454622906707
5527939700884757
8944394323791464
14472334024676221
23416728348467685
37889062373143906
61305790721611591
99194853094755497
160500643816367088
259695496911122585
420196140727489673
679891637638612258
1100087778366101931
1779979416004714189
2880067194370816120
4660046610375530309
754011380474634642

#### Przykłady:

##### Odczytywanie dużego pliku linia po linii
Załóżmy, że mamy bardzo duży plik i chcemy go odczytać linia po linii, zamiast ładować wszystko do pamięci na raz. Możemy do tego użyć yield.

Zamiast ładować cały plik do pamięci, każda linia jest przetwarzana osobno.

In [9]:
def read_large_file(file_name):
    with open(file_name, 'r') as file:
        for line in file:
            yield line.strip()

for line in read_large_file('large_file.txt'):
    print(line)



## read without yield
# def read_file(file_name):
#     with open(file_name, 'r') as file:
#         return file.readlines()

# content = read_file('large_file.txt')
# print(content)


Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean id massa augue. Praesent vehicula enim erat, vitae cursus ipsum eleifend ut. Morbi luctus metus sit amet ex iaculis, elementum viverra neque elementum. Nam quis sem pellentesque, mollis orci ac, dictum augue. Maecenas neque quam, dapibus faucibus mauris nec, scelerisque condimentum metus. Proin ultrices massa at dictum dapibus. Maecenas elementum ut elit ac ultricies. Integer risus lectus, volutpat tincidunt pulvinar eu, posuere ac odio. Nam congue nibh quis elit cursus tincidunt. Vestibulum ut urna a turpis feugiat volutpat a sit amet ex.

Aenean tincidunt ex nisl, sit amet vestibulum erat viverra eu. Duis eleifend massa quis ultrices vestibulum. Donec posuere porttitor lacinia. Pellentesque eget feugiat est, sed vestibulum velit. Nulla rhoncus ipsum et fringilla feugiat. Fusce luctus nulla et metus porta varius. Maecenas sit amet interdum magna. Quisque a finibus metus. Curabitur ultrices ultricies ante eget malesuada. I

### Ćwiczenia z List Comprehensions

1. **Stwórz listę kwadratów liczb od 1 do 10.**

   Oczekiwany wynik:
   `[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]`

2. **Stwórz listę parzystych liczb od 0 do 20.**

   Oczekiwany wynik:
   `[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20]`

3. **Zamień wszystkie litery w liście `['a', 'b', 'c', 'd']` na wielkie litery.**

   Oczekiwany wynik:
   `['A', 'B', 'C', 'D']`

4. **Stwórz listę długości słów z listy `['apple', 'banana', 'cherry', 'date']`.**

   Oczekiwany wynik:
   `[5, 6, 6, 4]`

5. **Stwórz listę liczb podzielnych przez 3 z zakresu od 1 do 50.**

   Oczekiwany wynik:
   `[3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 45, 48]`

6. **Stwórz listę liczb będących iloczynem odpowiednich elementów z dwóch list: `[1, 2, 3]` i `[4, 5, 6]`.**

   Oczekiwany wynik:
   `[4, 10, 18]`

7. **Zamień listę liczb `[1, -2, 3, -4, 5]` na ich wartości bezwzględne.**

   Oczekiwany wynik:
  `[1, 2, 3, 4, 5]`

8. **Stwórz listę liczb od 1 do 100, ale zamień wszystkie liczby podzielne przez 5 na słowo "Buzz".**

   Oczekiwany wynik:
 `[1, 2, 3, 4, 'Buzz', 6, 7, 8, 9, 'Buzz', 11, 12, 13, 14, 'Buzz', 16, 17, 18, 19, 'Buzz', 21, 22, 23, 24, 'Buzz', 26, 27, 28, 29, 'Buzz', 31, 32, 33, 34, 'Buzz', 36, 37, 38, 39, 'Buzz', 41, 42, 43, 44, 'Buzz', 46, 47, 48, 49, 'Buzz', 51, 52, 53, 54, 'Buzz', 56, 57, 58, 59, 'Buzz', 61, 62, 63, 64, 'Buzz', 66, 67, 68, 69, 'Buzz', 71, 72, 73, 74, 'Buzz', 76, 77, 78, 79, 'Buzz', 81, 82, 83, 84, 'Buzz', 86, 87, 88, 89, 'Buzz', 91, 92, 93, 94, 'Buzz', 96, 97, 98, 99, 'Buzz']`

9. **Filtruj tylko te słowa z listy `['dog', 'cat', 'rabbit', 'mouse']`, które mają więcej niż 3 litery.**

   Oczekiwany wynik:
   `['rabbit', 'mouse']`

10. **Stwórz listę tupli (liczba, jej kwadrat) dla liczb od 1 do 10.**

    Oczekiwany wynik:
    `[(1, 1), (2, 4), (3, 9), (4, 16), (5, 25), (6, 36), (7, 49), (8, 64), (9, 81), (10, 100)]`

11. **Stwórz listę liter z ciągu znaków `"python"` tylko dla tych, które są samogłoskami.**

    Oczekiwany wynik:
    `['o']`

12. **Stwórz listę liczb od 0 do 9 podniesionych do sześcianu, ale tylko dla tych liczb, które są nieparzyste.**

    Oczekiwany wynik:
    `[1, 27, 125, 343, 729]`

13. **Stwórz listę liczb od 1 do 30, które są zarówno podzielne przez 3, jak i 5.**

    Oczekiwany wynik:
    `[15, 30]`

14. **Zamień wszystkie liczby ujemne z listy `[-1, -2, 3, 4, -5]` na 0, a dodatnie pozostaw bez zmian.**

    Oczekiwany wynik:
    `[0, 0, 3, 4, 0]`

15. **Stwórz listę liczb od 1 do 100, ale dla liczb podzielnych przez 3 podstaw "Fizz", a dla liczb podzielnych przez 5 "Buzz".**

    Oczekiwany wynik:
    `[1, 2, 'Fizz', 4, 'Buzz', 'Fizz', 7, 8, 'Fizz', 'Buzz', 11, 'Fizz', 13, 14, 'FizzBuzz', 16, 17, 'Fizz', 19, 'Buzz', 'Fizz', 22, 23, 'Fizz', 'Buzz', 26, 'Fizz', 28, 29, 'FizzBuzz', 31, 32, 'Fizz', 34, 'Buzz', 'Fizz', 37, 38, 'Fizz', 'Buzz', 41, 'Fizz', 43, 44, 'FizzBuzz', 46, 47, 'Fizz', 49, 'Buzz', 'Fizz', 52, 53, 'Fizz', 'Buzz', 56, 57, 58, 59, 'FizzBuzz', 61, 62, 'Fizz', 64, 'Buzz', 'Fizz', 67, 68, 'Fizz', 'Buzz', 71, 72, 73, 74, 'FizzBuzz', 76, 77, 'Fizz', 79, 'Buzz', 'Fizz', 82, 83, 'Fizz', 'Buzz', 86, 87, 88, 89, 'FizzBuzz', 91, 92, 'Fizz', 94, 'Buzz', 'Fizz', 97, 98, 'Fizz', 'Buzz']`

16. **Utwórz listę pierwszych liter słów z listy `['Alice', 'Bob', 'Charlie', 'David']`.**

    Oczekiwany wynik:
    `['A', 'B', 'C', 'D']`

17. **Stwórz listę wszystkich liczb od 1 do 100, które są liczbami pierwszymi.**

    Oczekiwany wynik:
    `[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]`

18. **Utwórz listę odwrotnych postaci słów z listy `['hello', 'world', 'python']`.**

    Oczekiwany wynik:
    `['olleh', 'dlrow', 'nohtyp']`

19. **Wygeneruj macierz kwadratową o rozmiarze 3x3 wypełnioną liczbami od 1 do 9.**

    Oczekiwany wynik:
    `[[1, 2, 3], [4, 5, 6], [7, 8, 9]]`

20. **Utwórz listę liczb od 1 do 50, ale dla liczb podzielnych przez 2 zwróć "even", a dla pozostałych zwróć "odd".**

    Oczekiwany wynik:
    `['odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd', 'even', 'odd']`


### Ćwiczenia z Dictionary Comprehensions

1. **Stwórz słownik, w którym klucze to liczby od 1 do 5, a wartości to ich kwadraty.**

   Oczekiwany wynik:
   `{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}`

2. **Stwórz słownik, w którym klucze to litery z napisu `"hello"` (unikalne litery), a wartości to liczba wystąpień każdej litery w tym napisie.**

   Oczekiwany wynik:
   `{'h': 1, 'e': 1, 'l': 2, 'o': 1}`

3. **Stwórz słownik, w którym klucze to imiona z listy `['Alice', 'Bob', 'Charlie']`, a wartości to długości tych imion.**

   Oczekiwany wynik:
   `{'Alice': 5, 'Bob': 3, 'Charlie': 7}`

4. **Stwórz słownik, w którym klucze to liczby od 1 do 5, a wartości to liczby podzielne przez 2.**

   Oczekiwany wynik:
   `{1: 0, 2: 2, 3: 0, 4: 4, 5: 0}`

5. **Stwórz słownik, w którym klucze to liczby od 1 do 10, a wartości to ich potęgi trzeciej.**

   Oczekiwany wynik:
   `{1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216, 7: 343, 8: 512, 9: 729, 10: 1000}`


6. **Stwórz słownik, w którym klucze to słowa z listy `['apple', 'banana', 'cherry']`, a wartości to długości tych słów.**

    Oczekiwany wynik:
    `{'apple': 5, 'banana': 6, 'cherry': 6}`

7. **Stwórz słownik, w którym klucze to litery z napisu `"world"`, a wartości to ich pozycje w alfabecie (a=1, b=2, itd.).**

    Oczekiwany wynik:
    `{'w': 23, 'o': 15, 'r': 18, 'l': 12, 'd': 4}`


8. **Stwórz słownik, w którym klucze to liczby od 1 do 5, a wartości to ich liczby pierwsze (jeśli liczba jest pierwsza, to wartość jest równa kluczowi, w przeciwnym razie `None`).**

    Oczekiwany wynik:
    `{1: None, 2: 2, 3: 3, 4: None, 5: 5}`

9. **Stwórz słownik, w którym klucze to litery z napisu `"python"`, a wartości to liczby ich wystąpień w napisie.**

    Oczekiwany wynik:
    `{'p': 1, 'y': 1, 't': 1, 'h': 1, 'o': 1, 'n': 1}`


## Lambda, map, filter

**Funkcje lambda** to anonimowe funkcje, które są tworzone za pomocą słowa kluczowego `lambda`. Służą do definiowania funkcji w jednym wierszu i są często używane tam, gdzie funkcja jest przekazywana jako argument do innej funkcji.

**Składnia:**

```python
lambda arguments: expression
```

**Przykłady:**

In [2]:
# Lambda, która dodaje 10 do argumentu
add_ten = lambda x: x + 10
print(add_ten(5))


# Funkcja mnożąca dwa argumenty:
multiply = lambda x, y: x * y
print(multiply(4, 5))

# Funkcja zwracająca największą wartość z dwóch argumentów
max_value = lambda x, y: x if x > y else y
print(max_value(7, 10))

is_even = lambda x: x % 2 == 0
print(is_even(4))
print(is_even(7))


15
20


In [3]:
# Funkcja sortująca listę krotek według drugiego elementu krotki:

tuples = [(1, 3), (2, 2), (4, 1)]
sorted_tuples = sorted(tuples, key=lambda x: x[1])
print(sorted_tuples)  # Wyjście: [(4, 1), (2, 2), (1, 3)]


[(4, 1), (2, 2), (1, 3)]


## map()

Funkcja `map()` w Pythonie jest używana do stosowania danej funkcji do każdego elementu iterowalnego (takiego jak lista, krotka, czy inny obiekt iterowalny). `map()` zwraca iterator, który można przekonwertować na listę lub inny typ kolekcji.

```python
map(function, iterable, ...)

```

gdzie

- `function`: Funkcja, która ma być zastosowana do każdego elementu iterowalnego.
- `iterable`: Iterowalny obiekt, na którym będzie stosowana funkcja.


Jak to działa?
- Przygotowanie funkcji: Funkcja, którą przekazujesz do `map()`, jest stosowana do każdego elementu w iterowalnym obiekcie.
- Iteracja: `map()` przechodzi przez każdy element iterowalnego obiektu, stosując funkcję do każdego z nich.
- Zwracanie iteratora: `map()` zwraca iterator, który można przekształcić w listę, krotkę lub inny typ kolekcji.

#### Przykłady:

In [4]:
def square(x):
    return x ** 2

numbers = [1, 2, 3, 4, 5]
squared_numbers = map(square, numbers)
print(list(squared_numbers))


[1, 4, 9, 16, 25]


In [6]:
def add(x, y):
    return x + y

numbers1 = [1, 2, 3]
numbers2 = [4, 5, 6]
summed_numbers = map(add, numbers1, numbers2)
print(list(summed_numbers))

[5, 7, 9]


W tym przypadku `map()` stosuje funkcję add do odpowiadających sobie elementów z dwóch list. Funkcja add przyjmuje dwa argumenty i dodaje je do siebie.

## filter()

Funkcja `filter()` w Pythonie jest używana do filtrowania elementów iterowalnego obiektu (takiego jak lista, krotka, czy inny obiekt iterowalny) na podstawie funkcji, która zwraca wartość boolowską (`True` lub `False`). `filter()` zwraca iterator, który zawiera tylko te elementy, dla których funkcja zwróciła `True`.



```python
filter(function, iterable)

```

gdzie

- `function`: Funkcja, która stosuje warunek do każdego elementu iterowalnego. Powinna zwracać `True` lub `False`.
- `iterable`: Iterowalny obiekt, na którym będzie stosowana funkcja.


Jak to działa?
- Przygotowanie funkcji: Funkcja, którą przekazujesz do `filter()`, jest stosowana do każdego elementu w iterowalnym obiekcie. Funkcja ta musi zwracać `True` lub `False`.
- Iteracja: `filter()` przechodzi przez każdy element iterowalnego obiektu, stosując funkcję do każdego z nich.
- Zwracanie iteratora: `filter()` zwraca iterator, który zawiera tylko te elementy, dla których funkcja zwróciła True.

### Przykłady:

In [7]:
# Funkcja sprawdzająca, czy liczba jest parzysta
def is_even(x):
    return x % 2 == 0

numbers = [1, 2, 3, 4, 5, 6]
even_numbers = filter(is_even, numbers)
print(list(even_numbers))


[2, 4, 6]


In [8]:
# użycie filter() z lambda
numbers = [1, 2, 3, 4, 5, 6]
even_numbers = filter(lambda x: x % 2 == 0, numbers)
print(list(even_numbers))  # Wyjście: [2, 4, 6]


[2, 4, 6]


In [10]:
# Funkcja sprawdzająca, czy napis ma długość większą niż 3
def longer_than_three(s):
    return len(s) > 3

words = ["cat", "dog", "elephant", "fish"]
long_words = filter(longer_than_three, words)
print(list(long_words))  # Wyjście: ['elephant']




# z lambda
words = ["cat", "dog", "elephant", "fish"]
long_words = filter(lambda s: len(s) > 3, words)
print(list(long_words))  # Wyjście: ['elephant']


['elephant', 'fish']
['elephant', 'fish']


### Ćwiczenia:

### Ćwiczenia z map()

1. **Stwórz listę liczb od 1 do 10. Użyj funkcji `map()` do podniesienia każdej liczby do potęgi czwartej.**

   Oczekiwany wynik:
   `[1, 16, 81, 256, 625, 1296, 2401, 4096, 6561, 10000]`

2. **Stwórz listę napisów: `['hello', 'world', 'python']`. Użyj funkcji `map()` do zamiany każdego napisu na jego długość.**

   Oczekiwany wynik:
   `[5, 5, 6]`

3. **Stwórz listę liczb: `[10, 20, 30, 40, 50]`. Użyj funkcji `map()` do obliczenia pierwiastka kwadratowego z każdej liczby.**

   Oczekiwany wynik:
   `[3.1622776601683795, 4.47213595499958, 5.477225575051661, 6.324555320336759, 7.0710678118654755]`

4. **Stwórz listę liczb: `[2, 4, 6, 8]`. Użyj funkcji `map()` do konwersji każdej liczby na napis w formacie "Liczba: X", gdzie X to wartość liczby.**

   Oczekiwany wynik:
   `['Liczba: 2', 'Liczba: 4', 'Liczba: 6', 'Liczba: 8']`

### Ćwiczenia z filter()

1. **Stwórz listę liczb od 1 do 20. Użyj funkcji `filter()` do wybrania tylko tych liczb, które są większe niż 10.**

   Oczekiwany wynik:
   `[11, 12, 13, 14, 15, 16, 17, 18, 19, 20]`

2. **Stwórz listę napisów: `['cat', 'dog', 'elephant', 'fish']`. Użyj funkcji `filter()` do wybrania tylko tych napisów, które zawierają literę 'e'.**

   Oczekiwany wynik:
   `['elephant']`

3. **Stwórz listę liczb: `[12, 15, 22, 29, 35]`. Użyj funkcji `filter()` do wybrania tylko tych liczb, które są podzielne przez 5.**

   Oczekiwany wynik:
   `[15, 35]`

4. **Stwórz listę napisów: `['apple', 'banana', 'cherry', 'date']`. Użyj funkcji `filter()` do wybrania tylko tych napisów, które zaczynają się na literę 'b' lub 'c'.**

   Oczekiwany wynik:
   `['banana', 'cherry']`
